src/pages/VeSDL/OnChainVote.tsx

```diff
@@ -0,0 +1,247 @@
+  veSdlBalance: BigNumber
+  gauges: Partial<{
+    [lpTokenAddress: string]: Gauge
+  }>
+  gaugeControllerContract: GaugeController
+}
+export type GaugeName = {
+  [gaugeAddress: string]: {
+    address: string
+    gaugeName: string
+  }
+}
+
+export type VotesType = {
+  [gaugeAddress: string]: {
+    gaugeName: string
+    weight: BigNumber
+    voteDate: number
+  }
+}
+
+export default function OnChainVote({
+  veSdlBalance,
+  gauges,
+  gaugeControllerContract,
+}: OnChainVoteProps) {
+  const theme = useTheme()
+  const { account, chainId } = useActiveWeb3React()
+  const { t } = useTranslation()
+  const [selectedGauge, setSelectedGauge] = useState<{
+    gaugeName: string
+    address: string
+  } | null>(null)
+  const [alertMessage, setAlertMessage] = useState<string>()
+  const [voteWeightToSubmit, setVoteWeightToSubmit] = useState<string>("")
+  const [voteWeightInputError, setVoteWeightInputError] = useState<{
+    hasError: boolean
+    errorText: string
+  }>({ hasError: false, errorText: " " })
+
+  const gaugeNames = useMemo(() => {
+    const gaugeNameObj = {} as GaugeName
+    Object.keys(gauges)
+      .filter((lpTokenAddress) => !gauges[lpTokenAddress]?.isKilled)
+      .forEach((lpTokenAddress) => {
+        const gaugeName = gauges[lpTokenAddress]?.gaugeName
+        const gaugeAddress = gauges[lpTokenAddress]?.address.toLowerCase()
+        if (gaugeName && gaugeAddress)
+          gaugeNameObj[gaugeAddress] = {
+            address: gaugeAddress,
+            gaugeName: gaugeName,
+          }
+      })
+    return gaugeNameObj
+  }, [gauges])
+
+  const getVoteUsed = () => {
+    if (!account) return
+    return gaugeControllerContract["vote_user_power(address)"](account)
+  }
+  const getFilteredVotes = () => {
+    const voteFilter = gaugeControllerContract.filters.VoteForGauge(
+      null,
+      null,
+      null,
+      null,
+    )
+    return gaugeControllerContract.queryFilter(voteFilter)
+  }
+
+  const { data: voteUsed } = useQuery(["voteUsed"], getVoteUsed)
+  const { data: filteredRes } = useQuery(["filteredVotes"], getFilteredVotes)
+
+  const filteredVote = filteredRes?.reduce(
+    (acc, { args: { user, time, weight, gauge_addr } }) => {
+      if (user === account) {
+        acc[gauge_addr.toLowerCase()] = {
+          gaugeName: gaugeNames[gauge_addr.toLowerCase()].gaugeName,
+          voteDate: time.toNumber(),
+          weight,
+        }
+      }
+      return acc
+    },
+    {} as VotesType,
+  )
+
+  const handleVote = async () => {
+    if (voteWeightToSubmit && isNumberOrEmpty(voteWeightToSubmit) && chainId) {
+      try {
+        if (veSdlBalance.isZero()) {
+          setAlertMessage("You need veSDL to vote")
+        } else if (selectedGauge?.address && gaugeControllerContract) {
+          const txn = await gaugeControllerContract.vote_for_gauge_weights(
+            selectedGauge.address,
+            Math.round(parseFloat(voteWeightToSubmit) * 100), //convert percent to bps
```

Comment:
Using `Math.round` may lead to the txn not being able to execute as expected. If the user has `n`% voting power remaining, and we round their input up to `n + 1`, then they would appear to be voting with too much power.

Reply:
Fixed.